core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -669,6 +676,42 @@ private PlanWithProperties visitPartitionedWriter(PlanNode node, PartitioningSch
             return rebaseAndDeriveProperties(node, ImmutableList.of(exchange));
         }
 
+        private PlanWithProperties visitScalePartitionedWriter(PlanNode node, PartitioningScheme partitioningScheme, PlanNode source, StreamPreferredProperties parentPreferences)
+        {
+            if (partitioningScheme.getPartitioning().getHandle().equals(FIXED_HASH_DISTRIBUTION)) {
+                // arbitrary hash function on predefined set of partition columns
+                StreamPreferredProperties preference = partitionedOn(partitioningScheme.getPartitioning().getColumns());
+                PlanWithProperties newSource = source.accept(this, preference);
+                PlanWithProperties exchange = deriveProperties(
+                        partitionedExchange(
+                                idAllocator.getNextId(),
+                                LOCAL,
+                                newSource.getNode(),
+                                partitioningScheme.withAlternativePartitioningHandle(SCALED_WRITER_HASH_DISTRIBUTION)),
+                        newSource.getProperties());
+                return rebaseAndDeriveProperties(node, ImmutableList.of(exchange));
+            }
+
+            // connector provided hash function
+            verify(!(partitioningScheme.getPartitioning().getHandle().getConnectorHandle() instanceof SystemPartitioningHandle));
+            verify(
+                    partitioningScheme.getPartitioning().getArguments().stream().noneMatch(Partitioning.ArgumentBinding::isConstant),
+                    "Table writer partitioning has constant arguments");
+
+            PlanWithProperties newSource = source.accept(this, parentPreferences);
```

Comment:
why don't we prefer `StreamPreferredProperties preference = partitionedOn(partitioningScheme.getPartitioning().getColumns());` as for `FIXED_HASH_DISTRIBUTION`?